### ISSUE TITLE
Move `src/parser.c` to git LFS

### ISSUE THREAD

**Issue by tamasvajk** [tree-sitter/tree-sitter-c-sharp#273]
I started receiving the below warning on pushes:
```
remote: warning: See http://git.io/iEPt8g for more information.
remote: warning: File src/parser.c is 50.90 MB; this is larger than GitHub's recommended maximum file size of 50.00 MB
remote: warning: GH001: Large files detected. You may want to try Git Large File Storage - https://git-lfs.github.com.
```

I'm not sure what would be the impact of moving to LFS. Maybe some documentation would need to be updated.

**Comment by Sjord** [comment]
We could reconsider whether parser.c should be committed at all. It is auto-generated with tree-sitter generate, so I am not sure it should be in the repository.

**Comment by tamasvajk** [comment]
Not committing it would be even better. But I was under the impression that this is a de facto requirement of a tree sitter grammar repo, because this way all you have to do is check out the grammar’s git repository into a well-known location[^1]. I'm not sure if this is a hard requirement or if we can deviate from it.

[^1]: https://dcreager.net/tree-sitter/getting-started/#installing-a-grammar

**Comment by damieng** [comment]
I don't know enough about how tree-sitter resolves these precedence rules to be helpful here unfortunately (every time I read the docs I feel like I've got a good handle on it but then once I dig into our non-trivial rule set it escapes me)

Whatever we do here it might be worth adding some kind of file size echo to a text file so that we can see at a glance on PR's how a change affects large generated file sizes.

**Comment by sogaiu** [comment]
Re: `parser.c` and friends being part of the repository

I think nvim-treesitter and Emacs both have logic in them to compile shared libraries for parsers from files that live under `src/parser.c` and friends fetched from grammar repositories.

In nvim-treesitter's case I think there may be logic to call `tree-sitter generate` [under certain circumstances](https://github.com/nvim-treesitter/nvim-treesitter/blob/master/lua/nvim-treesitter/install.lua#L326-L346) but that means a user needs to have the `tree-sitter` cli installed IIUC.  Not sure of the specifics but I believe @theHamsta knows.

In Emacs 29+'s case I don't think there is any logic ATM to run the `tree-sitter`'s `generate` subcommand: https://github.com/emacs-mirror/emacs/blob/f4f30ff4c44dcfdf780f1981aa541af713f2805f/lisp/treesit.el#L2684-L2696

**Comment by sogaiu** [comment]
It looks like difftastic [uses tree-sitter-csharp](https://github.com/Wilfred/difftastic/blob/94b6117723e2416899d6aa9917e895ff8dd13c11/build.rs#L108-L112) and assumes [`src/parser.c` is available](https://github.com/Wilfred/difftastic/blob/94b6117723e2416899d6aa9917e895ff8dd13c11/build.rs#L28).

Not really sure though...but I would be surprised if @Wilfred didn't know :)

**Comment by sogaiu** [comment]
I'm not sure how cursorless does its building, but I think it also transitively(?) [uses tree-sitter-c-sharp](https://github.com/cursorless-dev/vscode-parse-tree/blob/0ad3949ba1b4d1e428b4da0fa67c63bd6ec87ba4/package.json#L91).

There's a [custom `web-tree-sitter` mentioned](https://github.com/cursorless-dev/vscode-parse-tree/blob/0ad3949ba1b4d1e428b4da0fa67c63bd6ec87ba4/package.json#L125) so may be Emscripten is involved in compiling `parser.c` and friends to produce `.wasm`.

May be @pokey knows the details.

**Comment by pokey** [comment]
Here's where we build the wasm:

https://github.com/cursorless-dev/vscode-parse-tree/blob/main/Makefile#L43

So, as long as `parser.c` is in the npm package, I think we're ok.  Fwiw we do intend to migrate all our tree-sitter language dependencies to use GitHub commit sha's instead of npm versions, because the npm versions tend to be stale.  I believe that means we'd run into some issues if you stopped including parser.c in the repo

That being said, I think we could probably add a step to generate parser.c before we compile the wasm, so should be ok.

But I do believe every other tree-sitter grammar repo keeps the generated `parser.c` in source control, so it may come as a surprise to other downstream consumers if it does not

**Comment by damieng** [comment]
I think this should be a discussion across tree-sitter grammars in general. Having tree-sitter-c-sharp be different/unusual from other grammars in this respect will not help adoption.

Unfortunately we don't really have a broad tree-sitter discussion channel. Perhaps we should start one.

I've create a Discord one if anyone is interested https://discord.gg/KrRa5ATb

**Comment by pokey** [comment]
Agreed.  A lot of discussion seems to happen on https://github.com/tree-sitter/tree-sitter/discussions (in particular https://github.com/tree-sitter/tree-sitter/discussions/1768 was a great one), but agreed a Discord server could be useful.  I just asked to see if there already was one (https://github.com/tree-sitter/tree-sitter/discussions/2027)

**Comment by theHamsta** [comment]
![grafik](https://user-images.githubusercontent.com/7189118/212476891-6ed231dc-fd15-479f-bf1c-c2f316269ff8.png)

There is an option to include Git LFS in archives, which would maybe a possibility for everyone using tree-sitter-c-sharp via tarball. When anyone is using Git, the Git LFS requirement would propagate to the consuming repo.

**Comment by tamasvajk** [comment]
I found this related discussion: https://github.com/tree-sitter/tree-sitter/discussions/1243.

**Comment by sogaiu** [comment]
I tried `tree-sitter generate` for tree-sitter-c-sharp and it really takes quite while doesn't it:
```
$ time tree-sitter generate

real	4m2.854s
user	3m52.457s
sys	0m10.372s
```

Memory usage exceeded 50GB here too...

**Comment by damieng** [comment]
Yes, it has grown quite considerably lately. We're going to have to make some trade-offs between handling every single scenario possible in C# vs performance I suspect.

**Comment by sogaiu** [comment]
Somewhat related, I noticed this issue at the tree-sitter-swift repository: https://github.com/alex-pinkus/tree-sitter-swift/issues/149

**Comment by damieng** [comment]
That's a pretty interesting solution, I'd be up for doing that here. I wonder if we can proactively reach out to dependencies we know will be affected.

**Comment by sogaiu** [comment]
The issue of it being non-trivial to contact users of one's grammar seems to be the sort of thing that many grammars will / do face.

I wonder if there is a good approach to this.

We started a sticky issue where we have began to announce potential upcoming changes and asked the users we know about to subscribe to it and let us know if things we're planning on could be a problem: https://github.com/sogaiu/tree-sitter-clojure/issues/33

We tried to ascertain who was using us and then went around to the folks we came up with, but later it turned out there were other folks.

From the perspective of a "user" (e.g. nvim-treesitter, Emacs 29+, difftastic, helix-editor, etc.) though, I would imagine that there would be a preference to not use have different methods of being up-to-date about upcoming changes.  Some kind of unified way seems like a better deal.

I wonder if there's something that can be done at the tree-sitter repository...like a dedicated thread(?) / discussion per grammar?

**Comment by damieng** [comment]
I wonder then if we flip what tree-sitter-swift did on it's head and instead move development to a new 'dev' branch that PR's need to be made against that has no generated files.

We'd then adopt the tree-sitter-swift approach that when a release is tagged we generate the files and merge them into `master` (rather than `with-generated-files`)

That way everyone who is relying on master and the generated artefacts today is unaffected other than lagging behind releases a bit. 

Given there's only a small handful of contributors to the repo the work in us changing from developing against master to developing against `dev` or `develop` would be quite small.

**Comment by Thaodan** [comment]
Why not the other way around? Unless the user knows they can expect generated files to be inside the repository they don't expect it. Also updating a generated file adds major noise to the commit history.

**Comment by Thaodan** [comment]
Improving the memory used to generate this grammar would be great. I'm at 26.5GB right now for some architectures I don't even have the memory to generate the grammar (Arm and PPC).

**Comment by damieng** [comment]
We'd love to be able to improve the memory usage but the only options I'm aware of there are either:

1. Reduce the amount of C# syntax we support
2. Make changes to the core tree-sitter engine

We really don't want to do option 1 as it would be a breaking change and make this library less useful. As for option 2 I wouldn't know where to start and I suspect the only people who do are really busy on other things.

**Comment by Thaodan** [comment]
I get the point, I wanted to highlight the amount of memory required to generate the current grammar.

If it only would be possible to generate parts of the grammar in a separate process.

**Comment by damieng** [comment]
It's definitely something that needs to be done but that would happen in tree-sitter itself not the C# grammar.

**Comment by ObserverOfTime** [comment]
> Maybe one could shake things about a little, like having tagged releases, and have the generated parser.c files part of the release artefacts? But that would require that someone sets up a GitHub Actions pipeline, and that it has enough **omph** to process the current grammars.

https://github.com/tree-sitter/workflows#release-workflow
